---
title: "Finn_UK_analysis"
output: html_notebook
---

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
text_size = 20
```

```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(VennDiagram)
library(reshape2)
library(vioplot)
library(viridis)
library(egg)
library(ggpubr)
library(ggupset)

setwd("/media/MIRROR/ukb_finngen/uk_finn_comparison/results/")

# mean normalization for betas
normalit<-function(m){
   (m - mean(m))/(max(m)-min(m))
 }
```

Load SNP table
```{r message=FALSE, warning=FALSE}
# SNP_data <- read_tsv("/media/MIRROR/ukb_finngen/uk_finn_comparison/results/SNP_table.tsv", col_types = "cnccncicciccicccccccc")
SNP_data <- read_tsv("/media/MIRROR/ukb_finngen/uk_finn_comparison/results/filtered_SNP_table.tsv", col_types = "cnccnnnncicciccicccccccc") %>% 
  filter(is_index == "index") %>% 
  filter(abs(beta_finn) < 1.2) %>%
  filter(abs(beta_UK) < 1.2) %>%
  mutate(across(c(beta_finn, beta_UK), normalit))

nominal_data <- read_tsv("/media/MIRROR/ukb_finngen/uk_finn_comparison/results/nominal_sign_table.tsv", col_types = "cnccnccncn", col_names = c("varid", "Finn_MAF", "Finn_rsid", "UKB_rsid", "UKB_MAF", "meta_listed", "nominal_phenonames_finn", "nom_pleio_finn", "nominal_phenonames_uk", "nom_pleio_uk"))

SNP_data <- SNP_data %>% 
  left_join(nominal_data, by = c("varid", "Finn_MAF", "Finn_rsid", "UKB_rsid", "UKB_MAF")) %>% 
  filter(meta_listed == "yes")

```
Load pheno table
```{r message=FALSE, warning=FALSE}
trait_data <- read_tsv("updated_table.tsv")


intersection_data <- read_tsv("clump_intersection_table.tsv")

new_clump_data <- trait_data %>% 
  select(Finn_code, UK_code, finn_clump_n, uk_clump_n, meta_clump_n) %>% 
  left_join(intersection_data, by = c("Finn_code", "UK_code")) %>% 
  select(Finn_code, UK_code, contains("clump"))

```
per trait boxplot SNP
```{r}
fig1_1 <- trait_data %>% 
  dplyr::select(-c("comments", "Unnamed: 8", "Unnamed: 10", "Соответствие")) %>% 
  na.omit() %>% 
  filter(meta_SNP_n > 0, uk_SNP_n > 0, finn_SNP_n > 0) %>% 
  melt(measure = c("finn_SNP_n", "uk_SNP_n", "meta_SNP_n"), id = "Finn_code", variable = "SNP_source") %>% 
  ggplot(aes(SNP_source, value))+
  geom_boxplot()+
  theme(text = element_text(size = text_size))+
  theme_bw()+
  labs(title="Number of SNPs associated with phenotypes by association source",
        x ="Biobank", y = "Number of associated SNPs")+
  scale_x_discrete(labels = c("Finngen", "UKBB", "Meta-analysis"))

```
per trait boxplot clump
```{r}
fig1_2 <- trait_data %>% 
  dplyr::select(-c("comments", "Unnamed: 8", "Unnamed: 10", "Соответствие")) %>% 
  na.omit() %>% 
  filter(meta_clump_n > 0, uk_clump_n > 0, finn_clump_n > 0) %>% 
  melt(measure = c("finn_clump_n", "uk_clump_n", "meta_clump_n"), id = "Finn_code", variable = "SNP_source") %>% 
  ggplot(aes(SNP_source, value))+
  geom_boxplot()+
  theme_bw()+
  theme(text = element_text(size = text_size))+
  labs(title="Number of genomic regions associated with phenotypes by association source",
        x ="Biobank", y = "Number of associated regions")+
  scale_x_discrete(labels = c("Finngen", "UKBB", "Meta-analysis"))
```
shared SNP boxplot
```{r}
fig1_3 <- trait_data %>% 
  dplyr::select(-c("comments", "Unnamed: 8", "Unnamed: 10", "Соответствие")) %>% 
  na.omit() %>% 
  filter(shared_SNP_fu_n > 0, shared_SNP_fm_n > 0, shared_SNP_um_n > 0, shared_SNP_fum_n > 0) %>% 
  melt(measure = c("shared_SNP_fu_n", "shared_SNP_fm_n", "shared_SNP_um_n", "shared_SNP_fum_n"), id = "Finn_code", variable = "SNP_source") %>% 
  ggplot(aes(SNP_source, value))+
  geom_boxplot()+
  theme_bw()+
  theme(text = element_text(size = text_size))+
  labs(title="Number of shared SNPs associated by association source",
        x ="", y = "Number of associated regions")+
  scale_x_discrete(labels = c("Finngen and UKBB", "Finngen and meta-analysis", "UKBB and meta-analysis", "All three"))
```

hexes
```{r}
p1 <- trait_data %>% 
  dplyr::select(-c("comments", "Unnamed: 8", "Unnamed: 10", "Соответствие")) %>% 
  na.omit() %>% 
  filter(meta_clump_n > 0, uk_clump_n > 0, finn_clump_n > 0) %>% 
  ggplot(aes(finn_clump_n, uk_clump_n))+
  geom_hex()+
  theme_bw()+
  theme(text = element_text(size = text_size))+
  labs(x = "FinnGen associated regions", y = "UKB associated regions")

p2 <- trait_data %>% 
  dplyr::select(-c("comments", "Unnamed: 8", "Unnamed: 10", "Соответствие")) %>% 
  na.omit() %>% 
  filter(meta_clump_n > 0, uk_clump_n > 0, finn_clump_n > 0) %>% 
  ggplot(aes(meta_clump_n, uk_clump_n))+
  geom_hex()+
  theme_bw()+
  theme(text = element_text(size = text_size))+
  labs(x = "Meta-analysis associated regions", y = "UKB associated regions")
  
p3 <- trait_data %>% 
  dplyr::select(-c("comments", "Unnamed: 8", "Unnamed: 10", "Соответствие")) %>% 
  na.omit() %>% 
  filter(meta_clump_n > 0, uk_clump_n > 0, finn_clump_n > 0) %>% 
  ggplot(aes(meta_clump_n, finn_clump_n))+
  geom_hex()+
  theme_bw()+
  theme(text = element_text(size = text_size))+
  labs(x = "Meta-analysis associated regions", y = "FinnGen associated regions")
  
  
p4 <- trait_data %>% 
  dplyr::select(-c("comments", "Unnamed: 8", "Unnamed: 10", "Соответствие")) %>% 
  na.omit() %>% 
  filter(meta_SNP_n > 0, uk_SNP_n > 0, finn_SNP_n > 0) %>% 
  ggplot(aes(finn_SNP_n, uk_SNP_n))+
  geom_point()+
  stat_cor()+
  geom_smooth(method = "lm")+
  theme_bw()+
  theme(text = element_text(size = text_size))+
  labs(x = "FinnGen associated SNPs", y = "UKB associated SNPs")


p5 <- trait_data %>% 
  dplyr::select(-c("comments", "Unnamed: 8", "Unnamed: 10", "Соответствие")) %>% 
  na.omit() %>% 
  filter(meta_SNP_n > 0, uk_SNP_n > 0, finn_SNP_n > 0) %>% 
  ggplot(aes(finn_SNP_n, meta_SNP_n))+
  geom_point()+
  stat_cor()+
  geom_smooth(method = "lm")+
  theme_bw()+
  theme(text = element_text(size = text_size))+
  labs(x = "FinnGen associated SNPs", y = "Meta-analysis associated SNPs")


p6 <- trait_data %>% 
  dplyr::select(-c("comments", "Unnamed: 8", "Unnamed: 10", "Соответствие")) %>% 
  na.omit() %>% 
  filter(meta_SNP_n > 0, uk_SNP_n > 0, finn_SNP_n > 0) %>% 
  ggplot(aes(uk_SNP_n, meta_SNP_n))+
  geom_point()+
  stat_cor()+
  geom_smooth(method = "lm")+
  theme_bw()+
  theme(text = element_text(size = text_size))+
  labs(x = "UKB associated SNPs", y = "Meta-analysis associated SNPs")


fig1_5 <- ggarrange(p1, p2, p3)

fig1_5
```



```{r message=F, warning=F}
plot_battery <- function(df1, df2, col_palette, labels)
{

# labels = c(deparse(substitute(df1)), deparse(substitute(df2)))

df1$source <- labels[1]
df2$source <- labels[2]

df <- rbind(df1, df2)

UK_MAF_test <- df %>% 
  melt(measure = c("Finn_MAF", "UKB_MAF"), variable = "SNP_source") %>% 
  wilcox.test(value ~ source, data = ., subset = SNP_source == "UKB_MAF")

Finn_MAF_test <- df %>% 
  melt(measure = c("Finn_MAF", "UKB_MAF"), variable = "SNP_source") %>% 
  wilcox.test(value ~ source, data = ., subset = SNP_source == "Finn_MAF")

MAF_finn_p <- paste("Wilcox test p-value", format(Finn_MAF_test$p.value, digits = 3))
MAF_uk_p <- paste("Wilcox test p-value", format(UK_MAF_test$p.value, digits = 3))
    
MAF_plot <- df %>% 
  melt(measure = c("Finn_MAF", "UKB_MAF"), variable = "SNP_source") %>% 
  ggplot(aes(x = source, y = value, fill = source))+
  geom_violin(width=1)+
  geom_boxplot(width=0.1, color="grey", fill="white", outlier.shape = NA)+
  scale_fill_manual(values = col_palette, name = element_blank())+
  theme_bw()+
  theme(text = element_text(size = 20), axis.text.x=element_blank(), strip.background = element_blank())+
  labs(y = "MAF", x = "")+
  facet_wrap("SNP_source", labeller = labeller("SNP_source" = c("Finn_MAF" = paste("FinnGen MAF", MAF_finn_p, sep = "\n"),
                                        "UKB_MAF" = paste("UKB MAF", MAF_uk_p, sep = "\n"))))

MAF_diff_test <- df %>% 
  mutate(maf_diff = abs(UKB_MAF - Finn_MAF)/pmax(UKB_MAF, Finn_MAF)) %>% 
  wilcox.test(maf_diff ~ source, data = .)
  
MAF_diff_plot <- df %>% 
  mutate(maf_diff = abs(UKB_MAF - Finn_MAF)/pmax(UKB_MAF, Finn_MAF)) %>% 
  ggplot(aes(x = source, y = maf_diff, fill = source))+
  geom_violin(width=1)+
  geom_boxplot(width=0.1, color="grey", fill="white", outlier.shape = NA)+
  scale_fill_manual(values = col_palette, name = element_blank()) +
  labs(caption = paste("Wilcox test p-value", format(MAF_diff_test$p.value, digits = 3))) + 
  theme_bw()+
  theme(text = element_text(size = 20), axis.text.x=element_blank())+
  labs(y = "Difference in MAF", x = "")

pleio_plot <- df %>%  
  melt(measure = c("pleio_finn", "pleio_ukb", "pleio_meta"), variable = "SNP_source") %>% 
  ggplot(aes(x = source, y = value, fill = source))+
  geom_violin(width=1, outlier.shape = NA)+
  geom_boxplot(width=0.1, color="grey", fill="white", outlier.shape = NA)+
  scale_fill_manual(values = col_palette, name = element_blank())+
  theme_bw()+
  theme(text = element_text(size = 20), axis.text.x=element_blank(), strip.background = element_blank())+
  ylim(0, 30)+
  labs(x = "", y = "Pleiotropy")+
  scale_x_discrete(labels = c("FinnGen", "UKB", "Meta-analysis"))+
  facet_wrap("SNP_source", labeller = labeller("SNP_source" = c("pleio_finn" = "FinnGen",
                                        "pleio_ukb" = "UKB", 
                                        "pleio_meta" = "Meta-analysis")))

beta_test_uk <- df %>% 
  melt(id = c("varid", "source"), measure = c("beta_finn", "beta_UK"), variable = "SNP_source") %>% 
  wilcox.test(value ~ source, data = ., subset = SNP_source == "beta_UK")

beta_test_finn <- df %>% 
  melt(id = c("varid", "source"), measure = c("beta_finn", "beta_UK"), variable = "SNP_source") %>% 
  wilcox.test(value ~ source, data = ., subset = SNP_source == "beta_finn")

beta_uk_p <- paste("Wilcox test p-value", format(beta_test_uk$p.value, digits = 3))
beta_finn_p <- paste("Wilcox test p-value", format(beta_test_finn$p.value, digits = 3))

beta_plot <- df %>% 
  melt(id = c("varid", "source"), measure = c("beta_finn", "beta_UK"), variable = "SNP_source") %>% 
  ggplot(aes(x = source, y = value, fill = source))+
  geom_violin(width=1)+
  geom_boxplot(width=0.1, color="grey", fill="white", outlier.shape = NA)+
  scale_fill_manual(values = col_palette, name = element_blank())+
  theme_bw()+
  theme(text = element_text(size = 20),
        axis.text.x=element_blank(),
        strip.background = element_blank())+
  labs(x = "", y = "Effect size")+
  ylim(0, 1.1)+
  facet_wrap("SNP_source", labeller = labeller("SNP_source" = c("beta_finn" = paste("FinnGen effect size", beta_finn_p, sep = "\n"),
                                        "beta_UK" = paste("UKB effect size", beta_uk_p, sep = "\n"))))
  
LD_test <- df %>%
  wilcox.test(LD_score ~ source, data = .)

LD_plot <- df %>% 
  ggplot(aes(x = source, y = LD_score, fill = source))+
  geom_violin(width=1)+
  geom_boxplot(width=0.1, color="grey", fill="white", outlier.shape = NA)+
  scale_fill_manual(values = col_palette, name = element_blank())+
  labs(x = "", caption = paste("Wilcox test p-value", format(LD_test$p.value, digits = 3)), y = "LD score") + 
  theme_bw()+
  theme(text = element_text(size = 20), axis.text.x=element_blank())

return(list(MAF_plot, MAF_diff_plot, pleio_plot, beta_plot, LD_plot))
}
```


Fig 1
shared clump boxplot
```{r}
fig1_4 <- new_clump_data %>% 
  na.omit() %>%  
  mutate(across(c(intersect_clump_fm_n, intersect_clump_fu_n, intersect_clump_um_n), ~ . + intersect_clump_fum_n)) %>%
  mutate(`Meta-analysis` = meta_clump_n, UKB = uk_clump_n, FinnGen = finn_clump_n) %>% 
  mutate(FinnGen_UKB = intersect_clump_fu_n, `FinnGen_Meta-analysis` = intersect_clump_fm_n, `UKB_Meta-analysis` = intersect_clump_um_n, `FinnGen_UKB_Meta-analysis` = intersect_clump_fum_n) %>% 
  melt(measure = c("FinnGen_UKB", "FinnGen_Meta-analysis", "UKB_Meta-analysis", "FinnGen_UKB_Meta-analysis", "FinnGen", "UKB", "Meta-analysis"), id = "Finn_code", variable = "SNP_source") %>% 
  filter(value > 0) %>% 
  ggplot(aes(SNP_source, value))+
  geom_violin(width=1, outlier.shape = NA)+
  geom_boxplot(width=0.1, color="grey", fill="white", outlier.shape = NA)+
  axis_combmatrix(sep = "_")+
  theme_bw()+
  ylim(0, 50)+
  theme(text = element_text(size = 20))+
  labs(title="Number associated genomic regions by association source",
        x ="", y = "Number of associated regions")

fig1_4
```

```{r message=FALSE, warning=FALSE}
library(tidytext)
unnested_data <- SNP_data %>% 
  tidytext::unnest_tokens(phenonames_finn, phenonames_finn, token = "regex", pattern = ",") %>% 
  tidytext::unnest_tokens(phenonames_ukb, phenonames_ukb, token = "regex", pattern = ",") %>% 
  tidytext::unnest_tokens(phenonames_meta, phenonames_meta, token = "regex", pattern = ",")

area1 = unnested_data %>% filter(pleio_finn > 0) %>% pull(varid)
area2 = unnested_data %>% filter(pleio_ukb > 0) %>% pull(varid)
area3 = unnested_data %>% filter(pleio_meta > 0) %>% pull(varid)


library(ggvenn)
# Anti-meta specific
unnested_data %>% 
  filter(pleio_finn > 0) %>% 
  filter(pleio_ukb > 0) %>% 
  filter(pleio_meta == 0) %>% 
  distinct(varid)

fig1_6 <- ggvenn(list(FinnGen = area1, UKB = area2, 'Meta-analysis' = area3))+
  theme(text = element_text(size = 10))

fig1_6
```

```{r}
pdf("fig1.pdf", 20, 10)

ggarrange(fig1_6, fig1_4)

dev.off()
```
Meta-specific
## meta-specific variants
```{r}
meta_SNPs <- SNP_data %>% 
  filter(pleio_finn == 0) %>% 
  filter(pleio_meta > 0) %>% 
  filter(pleio_ukb == 0)

# Distinct meta phenotypes
# unnested_data %>% 
#   filter(pleio_finn == 0) %>% 
#   filter(pleio_meta > 0) %>% 
#   filter(pleio_ukb == 0) %>% 
#   mutate(phenonames_meta = str_replace(phenonames_meta, pattern = ":.*", replacement = "")) %>% 
#   distinct(phenonames_meta)


non_meta_SNPs <- SNP_data %>% 
  filter(pleio_finn > 0 | pleio_ukb > 0) %>% 
  anti_join(meta_SNPs, by = "varid")
```

```{r}
meta_plots <- plot_battery(meta_SNPs, non_meta_SNPs, col_palette = c("yellow", "green"), labels = c("Meta-analysis specific SNPs", "Nonmeta-analysis specific SNPs"))

meta_plots[[1]]
meta_plots[[2]]
meta_plots[[3]]
meta_plots[[4]]
meta_plots[[5]]
```


```{r}
pdf("fig3.pdf", 20, 10)

ggarrange(meta_plots[[1]], meta_plots[[2]], meta_plots[[4]], meta_plots[[5]], common.legend = T, legend = "bottom")

dev.off()
```
Meta-reproducible
```{r}
finn_meta_repr_SNPs <- SNP_data %>% 
  filter(str_detect(repr_fm, "perfect|partial") == T)

uk_meta_repr_SNPs <- SNP_data %>% 
  filter(str_detect(repr_um, "perfect|partial") == T)

meta_repr_SNPs <- rbind(finn_meta_repr_SNPs, uk_meta_repr_SNPs) %>% 
  distinct()

meta_non_repr_SNPs <- SNP_data %>% 
  filter(str_detect(repr_fm, "none") == T) %>% 
  filter(str_detect(repr_um, "none") == T) %>% 
  anti_join(meta_SNPs, by = "varid")

meta_repr_plots <- plot_battery(meta_repr_SNPs, meta_non_repr_SNPs, c("magenta", "cyan"), labels = c("Reproducing variants", "Non-reproducing variants"))



# Mysterious SNPs
hmm1 <- meta_non_repr_SNPs %>% 
  filter(pleio_finn > 0) %>% 
  filter(pleio_meta == 0) %>% 
  filter(pleio_ukb > 0)


hmm2 <- meta_non_repr_SNPs %>% 
  filter(pleio_finn > 0) %>% 
  filter(pleio_meta == 0) %>% 
  filter(pleio_ukb == 0)


hmm3 <- meta_non_repr_SNPs %>% 
  filter(pleio_finn == 0) %>% 
  filter(pleio_meta > 0) %>% 
  filter(pleio_ukb == 0)

 
hmm4 <- meta_non_repr_SNPs %>% 
  filter(pleio_finn == 0) %>% 
  filter(pleio_meta == 0) %>% 
  filter(pleio_ukb > 0)

meta_non_repr_SNPs %>% 
  anti_join(rbind(hmm1, hmm2, hmm3, hmm4))



meta_repr_plots[[1]]
meta_repr_plots[[2]]
meta_repr_plots[[3]]
meta_repr_plots[[4]]
meta_repr_plots[[5]]
```
Fig 3
```{r}
pdf("fig2.pdf", 20, 10)

ggarrange(meta_repr_plots[[1]], meta_repr_plots[[2]], meta_repr_plots[[4]], meta_repr_plots[[5]], common.legend = T, legend = "bottom")

dev.off()
```
Supp cases
```{r}
finn_cohort <- read_tsv("../resources/Finn_cohort.tsv")
uk_cohort <- read_tsv("../resources/UKB_cohort.tsv")

pdf("Supp1.pdf", 20, 10)

trait_data %>% 
  select(Finn_code, UK_code, finn_clump_n) %>% 
  left_join(finn_cohort, by = c("Finn_code" = "phenocode")) %>% 
  rename(finn_cases = n_cases, finn_controls = n_controls) %>% 
  left_join(uk_cohort, by = c("UK_code" = "phenotype")) %>% 
  rename(uk_cases = n_cases, uk_controls = n_controls) %>% 
  ggplot(aes(finn_cases, uk_cases))+
  geom_point()+
  coord_fixed(clip = "off")+
  theme_bw()+
  theme(text = element_text(size = 30))+
  labs(y = "Cases in UKB", x = "Cases in FinnGen")+
  ggtitle("Cases number comparison across biobanks")

dev.off()
```
Supp cohort sizes
```{r}
cohort_sizes <- tibble(size = c(260405, 361194), biobank = c("FinnGen", "UKB"))

S1_plot <- cohort_sizes %>% 
ggplot(aes(biobank, size, fill = biobank))+
  geom_bar(stat = "identity")+
  theme_bw()+
  theme(text = element_text(size = 20))+
  labs(y = "Cohort size", x = "Biobank")


pdf("Supp2.pdf", 20, 10)

S1_plot

dev.off()

```
Supp hexes
```{r}
pdf("Supp3.pdf", 20, 15)

fig1_5

dev.off()
```
Supp pleio
```{r}
pleio_plot <- function(df1, df2, col_palette, labels)
{


df1$source <- labels[1]
df2$source <- labels[2]

df <- rbind(df1, df2)

pleio_plot <- df %>%  
  melt(measure = c("pleio_finn", "pleio_ukb", "pleio_meta"), variable = "SNP_source") %>% 
  mutate(is_pleio = value > 1) %>% 
  ggplot(aes(x = SNP_source, fill = is_pleio))+
  geom_bar(position = "fill", stat = "count")+
  scale_fill_brewer(guide = guide_legend(reverse=TRUE), palette = col_palette, name = "Is variant pleiotropic?", labels = c("no", "yes"))+
  theme_bw()+
  theme(text = element_text(size = 25),
        axis.title.x=element_blank())+
  
  
  labs(y = "Proportion of SNPs")+
  scale_x_discrete(labels = c("FinnGen", "UKB", "Meta-analysis"))+

  facet_wrap("source")

return(pleio_plot)
}

pleio_plot(repr_SNPs, non_repr_SNPs, c("red", "blue"), labels = c("Reproducing variants", "Non-reproducing variants"))

pdf("Supp4.pdf", 30, 10)

pleio1 <- pleio_plot(repr_SNPs, non_repr_SNPs, "Set1", labels = c("Reproducing variants", "Non-reproducing variants"))
pleio2 <- pleio_plot(meta_SNPs, non_meta_SNPs, "Set3", labels = c("Meta-analysis specific SNPs", "Nonmeta-analysis specific SNPs"))

ggarrange(pleio1, pleio2)

dev.off()

```
Nominal significance
```{r}
# Variants w/o GW significance in Finngen
UKB_GW <- SNP_data %>% 
  filter(pleio_finn == 0) %>% 
  filter(pleio_ukb > 0)

# Nominal significance and listed in our pairs
UKB_GW_paired <- UKB_GW %>% 
  transmute(varid, phenonames_ukb, nominal_phenonames_finn) %>% 
  tidytext::unnest_tokens(phenonames_ukb, phenonames_ukb, token = "regex", pattern = ",") %>% 
  tidytext::unnest_tokens(nominal_phenonames_finn, nominal_phenonames_finn, token = "regex", pattern = ",") %>% 
  mutate(nominal_phenonames_finn = str_replace(nominal_phenonames_finn, pattern = ":.*", replacement = "") %>% toupper(), 
         phenonames_ukb = str_replace(phenonames_ukb, pattern = ":.*", replacement = "") %>% toupper()) %>% 
  left_join(trait_data %>%
              transmute(UK_code, Finn_code, priority), 
            keep = T, 
            by = c("phenonames_ukb" = "UK_code", "nominal_phenonames_finn" = "Finn_code")) %>% 
  na.omit() %>% 
  distinct(varid, phenonames_ukb, nominal_phenonames_finn, .keep_all = T)

UKB_GW_paired_prior2 <- UKB_GW_paired %>% 
  filter(priority == 2)

ggvenn(list(Nominal = UKB_GW %>% pull(varid), Pairs = UKB_GW_paired %>% pull(varid), Second_priority = UKB_GW_paired_prior2 %>% pull(varid)))

# Variants w/o GW significance in UKB
Finn_GW <- SNP_data %>% 
  filter(pleio_finn > 0) %>% 
  filter(pleio_ukb == 0)

# Nominal significance and listed in our pairs
Finn_GW_paired <- Finn_GW %>% 
  transmute(varid, phenonames_finn, nominal_phenonames_uk) %>% 
  unnest_tokens(phenonames_finn, phenonames_finn, token = "regex", pattern = ",") %>% 
  unnest_tokens(nominal_phenonames_uk, nominal_phenonames_uk, token = "regex", pattern = ",") %>% 
  mutate(nominal_phenonames_uk = str_replace(nominal_phenonames_uk, pattern = ":.*", replacement = "") %>% toupper(), 
         phenonames_finn = str_replace(phenonames_finn, pattern = ":.*", replacement = "") %>% toupper()) %>% 
  left_join(trait_data %>%
              transmute(UK_code, Finn_code, priority), 
            keep = T, 
            by = c("nominal_phenonames_uk" = "UK_code", "phenonames_finn" = "Finn_code")) %>% 
  na.omit() %>% 
  distinct(varid, phenonames_finn, nominal_phenonames_uk, .keep_all = T)

Finn_GW_paired_prior2 <- Finn_GW_paired %>% 
  filter(priority == 2)

stacked_finn_data <- Finn_GW %>% 
  anti_join(Finn_GW_paired, by = 'varid') %>% 
  mutate(group = "Nominal") %>% 
  transmute(varid, group) %>% 
  rbind(Finn_GW_paired %>% 
          mutate(group = "Paired") %>% 
          transmute(varid, group)) %>% 
  anti_join(Finn_GW_paired_prior2, by = 'varid') %>%
  transmute(varid, group) %>% 
  rbind(Finn_GW_paired_prior2 %>% mutate(group = "Prior") %>% transmute(varid, group)) %>% 
  mutate(source = "Finngen")

stacked_uk_data <- UKB_GW %>% 
  anti_join(UKB_GW_paired, by = 'varid') %>% 
  mutate(group = "Nominal") %>% 
  transmute(varid, group) %>% 
  rbind(UKB_GW_paired %>% 
          mutate(group = "Paired") %>% 
          transmute(varid, group)) %>% 
  anti_join(UKB_GW_paired_prior2, by = 'varid') %>%
  transmute(varid, group) %>% 
  rbind(UKB_GW_paired_prior2 %>% mutate(group = "Prior") %>% transmute(varid, group)) %>% 
  mutate(source = "UKB")

  
stacked_finn_data %>%
  ggplot(aes(x = 1, alpha = group, fill = group))+
  scale_alpha_discrete(range = c(0.7, 0.8, 0.95))+
  geom_bar(position = 'stack')

stacked_uk_data %>%
  ggplot(aes(x = 1, alpha = group, fill = group))+
  scale_alpha_discrete(range = c(0.7, 0.8, 0.95))+
  geom_bar(position = 'stack')


pdf("Supp5.pdf", 30, 10)

rbind(stacked_uk_data, stacked_finn_data) %>% 
  ggplot(aes(x = source, fill = group))+
  scale_fill_brewer(palette = "Set1", name = "Group of SNPs", labels = c("Other traits", "Similar trait pairs", "Matching trait pairs"))+
  geom_bar(position = 'fill')+
  theme_bw()+
  theme(text = element_text(size = 30),
        axis.title.x=element_blank())+
  scale_x_discrete(labels = c("FinnGen", "UKB"))+
  ylab("Proportion of SNPs")

dev.off()
```
## Additional plots
GWAS catalog

```{r}
GWAS_catalog <- read_tsv("gwas_catalog_v1.0-associations_e109_r2023-06-22.tsv")

nonGWAScatalog <- SNP_data %>% 
  filter(pleio_meta > 0) %>% 
  anti_join(GWAS_catalog %>% 
              transmute(`DISEASE/TRAIT`, SNPS), by = c("Finn_rsid" = "SNPS"))


# Calculate phenotypes in novel findings
unnested_data %>% 
  filter(pleio_meta > 0) %>% 
  anti_join(GWAS_catalog %>% 
              transmute(`DISEASE/TRAIT`, SNPS), by = c("Finn_rsid" = "SNPS")) %>% 
  
  mutate(phenonames_meta = str_replace(phenonames_meta, pattern = ":.*", "")) %>% 
  distinct(phenonames_meta) %>% 
  na.omit

# Calculate chromosomes
unnested_data %>% 
  filter(pleio_meta > 0) %>% 
  anti_join(GWAS_catalog %>% 
              transmute(`DISEASE/TRAIT`, SNPS), by = c("Finn_rsid" = "SNPS")) %>% 
  
  mutate(chr = str_replace(varid, pattern = ":.*", "")) %>% 
  distinct(chr) %>% 
  na.omit

# Plot to check spread
gwas_data <- unnested_data %>% 
  filter(pleio_meta > 0) %>% 
  anti_join(GWAS_catalog %>% 
              transmute(`DISEASE/TRAIT`, SNPS), by = c("Finn_rsid" = "SNPS")) %>% 
  transmute(varid, pleio_meta) %>% 
  distinct(varid, .keep_all=T) %>% 
  separate(varid, into = c("chr", "bp", "ref", "alt")) %>% 
  mutate(chr = as.numeric(chr), bp = as.numeric(bp))

data_cum <- gwas_data %>% 
  group_by(chr) %>% 
  summarise(max_bp = max(bp)) %>% 
  mutate(bp_add = lag(cumsum(max_bp), default = 0)) %>% 
  select(chr, bp_add)

gwas_data <- gwas_data %>% 
  inner_join(data_cum, by = "chr") %>% 
  mutate(bp_cum = bp + bp_add)

axis_set <- gwas_data %>% 
  group_by(chr) %>% 
  summarize(center = mean(bp_cum))

ggplot(gwas_data, aes(x = bp_cum, y = pleio_meta, 
                                  color = as_factor(chr))) +
  geom_point(alpha = 0.75) +
  scale_x_continuous(label = axis_set$chr, breaks = axis_set$center) +
  scale_color_manual(values = rep(c("#276FBF", "#183059"), 
                                  unique(length(axis_set$chr)))) +
  
  labs(x = NULL, 
       y = "pleio_meta") + 
  theme_minimal() +
  theme( 
    legend.position = "none",
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    
    axis.text.x = element_text(angle = 60, size = 8, vjust = 0.5)
  )

meta_found <- SNP_data %>% 
  filter(pleio_meta > 0)

nonGWAScatalog %>% dim() %>% .[[1]] / meta_found %>% dim() %>% .[[1]]
```
Old repr
```{r}
repr_SNPs <- SNP_data %>% 
  filter(str_detect(repr_fu, "perfect|partial") == T)


perfect_repr_SNPs <- SNP_data %>% 
  filter(str_detect(repr_fu, "perfect") == T)

non_perfect_SNPs <- SNP_data %>% 
  filter(str_detect(repr_fu, "none|partial") == T)

non_repr_SNPs <- SNP_data %>% 
  filter(str_detect(repr_fu, "none") == T)
```
Repr - non-repr plots
```{r}
repr_plots <- plot_battery(repr_SNPs, non_repr_SNPs, c("red", "blue"), labels = c("Reproducing variants", "Non-reproducing variants"))

repr_plots[[1]]
repr_plots[[2]]
repr_plots[[3]]
repr_plots[[4]]
repr_plots[[5]]
```
Check interesting cases
```{r}
nonGWAScatalog %>% 
  inner_join(meta_SNPs, by = "varid") %>% 
  select(varid, Finn_MAF.x, UKB_MAF.x, phenonames_meta.x) %>% 
  View()
```


